### PR TITLE
Add Kotlin K2 migrations

### PR DESCRIPTION
**Pull request by tbogdanova**
Building the project with the K2 version of the compiler revealed several migrations. 
It is necessary to move these changes to `master` branch to simplify the support of `kotlin-community/dev` branch. As well as simplifying the future migration of the project to a new version of the compiler.

### CODE REVIEW

File: core/commonTest/src/contract/list/ImmutableListTest.kt [hunk 1]
```diff
@@ -25,7 +25,7 @@ class ImmutableListTest {
 
         assertFailsWith<NoSuchElementException> { empty1.iterator().next() }
 
-        compareLists(emptyList(), empty1)
+        compareLists(emptyList<Int>(), empty1)
```
Comment (ilya-g) [inline core/commonTest/src/contract/list/ImmutableListTest.kt:28]: AFAIK, this won't be a problem in K2 soon

Comment (tbogdanova) [reply core/commonTest/src/contract/list/ImmutableListTest.kt:28]: Warning will remain. See [KT-66196](https://youtrack.jetbrains.com/issue/KT-66196).

Comment (tbogdanova) [reply core/commonTest/src/contract/list/ImmutableListTest.kt:28]: But let's decide if we want to take this into the main branch.

Comment (qurbonzoda) [reply core/commonTest/src/contract/list/ImmutableListTest.kt:28]: Given that the change is needed for K2, I don't have objections to merging it into `master`. This also aligns with the goal of keeping the `kotlin-community/dev` branch close to `master`.

**Review by qurbonzoda — APPROVED** [review]

**Pull request merged**